lib/Conversion/FSMToCore/FSMToCore.cpp

```diff
@@ -0,0 +1,549 @@
+
+  OpBuilder::InsertionGuard guard(builder);
+  builder.setInsertionPointToStart(&region.front());
+
+  // Clone ConstantLike operations into the region.
+  for (auto &capture : captures) {
+    Operation *op = capture.getDefiningOp();
+    if (!op || !op->hasTrait<OpTrait::ConstantLike>())
+      continue;
+
+    Operation *cloned = builder.clone(*op);
+    for (auto [orig, replacement] :
+         llvm::zip(op->getResults(), cloned->getResults()))
+      replaceAllUsesInRegionWith(orig, replacement, region);
+  }
+}
+
+static ClkRstIdxs getMachinePortInfo(SmallVectorImpl<hw::PortInfo> &ports,
+                                     MachineOp machine, OpBuilder &b) {
+  // Get the port info of the machine inputs and outputs.
+  machine.getHWPortInfo(ports);
+  ClkRstIdxs specialPorts;
+
+  // Add clock port.
+  hw::PortInfo clock;
+  clock.name = b.getStringAttr("clk");
+  clock.dir = hw::ModulePort::Direction::Input;
+  clock.type = seq::ClockType::get(b.getContext());
+  clock.argNum = machine.getNumArguments();
+  ports.push_back(clock);
+  specialPorts.clockIdx = clock.argNum;
+
+  // Add reset port.
+  hw::PortInfo reset;
+  reset.name = b.getStringAttr("rst");
+  reset.dir = hw::ModulePort::Direction::Input;
+  reset.type = b.getI1Type();
+  reset.argNum = machine.getNumArguments() + 1;
+  ports.push_back(reset);
+  specialPorts.resetIdx = reset.argNum;
+
+  return specialPorts;
+}
+} // namespace
+
+namespace {
+
+class StateEncoding {
+
+public:
+  StateEncoding(OpBuilder &b, MachineOp machine, hw::HWModuleOp hwModule);
+
+  /// Get the encoded value for a state.
+  Value encode(StateOp state);
+  /// Get the state corresponding to an encoded value.
+  StateOp decode(Value value);
+
+  /// Returns the type which encodes the state values.
+  Type getStateType() { return stateType; }
+
+protected:
+  /// Creates a constant value in the module for the given encoded state
+  /// and records the state value in the mappings.
+  void setEncoding(StateOp state, Value v);
+
+  /// A mapping between a StateOp and its corresponding encoded value.
+  SmallDenseMap<StateOp, Value> stateToValue;
+
+  /// A mapping between an encoded value and its corresponding StateOp.
+  SmallDenseMap<Value, StateOp> valueToState;
+
+  Type stateType;
+
+  OpBuilder &b;
+  MachineOp machine;
+  hw::HWModuleOp hwModule;
+};
+
+StateEncoding::StateEncoding(OpBuilder &b, MachineOp machine,
+                             hw::HWModuleOp hwModule)
+    : b(b), machine(machine), hwModule(hwModule) {
+  Location loc = machine.getLoc();
+
+  OpBuilder::InsertionGuard guard(b);
+  b.setInsertionPointToStart(&hwModule.getBodyRegion().front());
+  // If stateType is explicitly provided, use this - otherwise, calculate the
+  // minimum int size that can represent all states
+  if (machine->getAttr("stateType")) {
+    // We already checked that a static cast is valid
+    stateType = cast<TypeAttr>(machine->getAttr("stateType")).getValue();
```

Comment:
What if the state type is too small to hold all state indices? Should we error out in that case?